### PR TITLE
Transit Gateway multicast support #11120 

### LINKED ISSUE #11120: Transit Gateway multicast support
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

AWS Transit Gateway now supports routing internet protocol (IP) multicast traffic between attached Amazon Virtual Private Cloud (VPC) connections.

### New or Affected Resource(s)

* `aws_ec2_transit_gateway_multicast_domain`

### Potential Terraform Configuration

```hcl
# Copy-paste your Terraform configurations here - for large Terraform configs,
# please use a service like Dropbox and share a link to the ZIP file. For
# security, you can also encrypt the files using our GPG public key.
```

### References

[Announcement](https://aws.amazon.com/about-aws/whats-new/2019/12/run-ip-multicast-workloads-aws-transit-gateway/).
[Blog post](https://aws.amazon.com/blogs/aws/aws-transit-gateway-adds-multicast-and-inter-regional-peering/).

Requires [AWS SDK v1.25.46](https://github.com/aws/aws-sdk-go/releases/tag/v1.25.46):
* https://github.com/terraform-providers/terraform-provider-aws/pull/11116

### LINKED ISSUE #16716: Transit Gateway IGMP support for multicast
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

> AWS Transit Gateway supports Internet Group Management Protocol (IGMP) for simplified deployment and management of multicast applications.

### New or Affected Resource(s)

* `aws_ec2_transit_gateway_multicast_domain`

### Potential Terraform Configuration

```hcl
# Copy-paste your Terraform configurations here - for large Terraform configs,
# please use a service like Dropbox and share a link to the ZIP file. For
# security, you can also encrypt the files using our GPG public key.
```

### References

Relates:

* https://github.com/hashicorp/terraform-provider-aws/issues/11120

[Announcement](https://aws.amazon.com/about-aws/whats-new/2020/12/aws-customers-use-industry-standard-internet-group-management-protocol/).
[User Guide](https://docs.aws.amazon.com/vpc/latest/tgw/tgw-multicast-overview.html).

Requires [AWS SDK v1.36.6](https://github.com/aws/aws-sdk-go/releases/tag/v1.36.6):

* https://github.com/hashicorp/terraform-provider-aws/pull/16713

### PR DESCRIPTION
**Pull request by NetDevAutomate**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

PR to add multicast support to Transit Gateway.
Closes #11120.
Closes #16716.

### CODE REVIEW

**Comment by NetDevAutomate** [comment]
Continued work from PR https://github.com/hashicorp/terraform-provider-aws/issues/11120

File: internal/service/ec2/transit_gateway_multicast_domain.go [hunk 1]
```diff
@@ -0,0 +1,788 @@
...
+			d.Id(), aws.StringValue(multicastDomain.State))
+		d.SetId("")
+		return nil
+	}
+
+	d.Set("igmpv2_support", multicastDomain.Options.Igmpv2Support)
+	d.Set("static_source_support", multicastDomain.Options.StaticSourcesSupport)
+	d.Set("auto_accept_shared_associations", multicastDomain.Options.AutoAcceptSharedAssociations)
+
+	// if err := d.Set("tags", ec2TagSpecificationsFromKeyValueTags(multicastDomain.Tags).IgnoreAws().IgnoreConfig(ignoreTagsConfig).Map()); err != nil {
```
Comment (drewmullen) [inline internal/service/ec2/transit_gateway_multicast_domain.go:215]: Can this be removed?

Comment (NetDevAutomate) [reply internal/service/ec2/transit_gateway_multicast_domain.go:215]: Done :-)

**Comment by NetDevAutomate** [comment]
$ make testacc TESTARGS='-run=TestAccEC2TransitGateway_serial/MulticastDomain' ACCTEST_PARALLELISM=5 PKG=ec2
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/ec2/... -v -count 1 -parallel 5  -run=TestAccEC2TransitGateway_serial/MulticastDomain -timeout 180m
=== RUN   TestAccEC2TransitGateway_serial
=== RUN   TestAccEC2TransitGateway_serial/MulticastDomain
=== RUN   TestAccEC2TransitGateway_serial/MulticastDomain/TransitGatewayMulticastDomainGroups
=== PAUSE TestAccEC2TransitGateway_serial/MulticastDomain/TransitGatewayMulticastDomainGroups
=== RUN   TestAccEC2TransitGateway_serial/MulticastDomain/basic
=== PAUSE TestAccEC2TransitGateway_serial/MulticastDomain/basic
=== RUN   TestAccEC2TransitGateway_serial/MulticastDomain/disappears
=== PAUSE TestAccEC2TransitGateway_serial/MulticastDomain/disappears
=== RUN   TestAccEC2TransitGateway_serial/MulticastDomain/Tags
=== PAUSE TestAccEC2TransitGateway_serial/MulticastDomain/Tags
=== RUN   TestAccEC2TransitGateway_serial/MulticastDomain/TransitGatewayMulticastDomainAssociation
=== PAUSE TestAccEC2TransitGateway_serial/MulticastDomain/TransitGatewayMulticastDomainAssociation
=== CONT  TestAccEC2TransitGateway_serial/MulticastDomain/TransitGatewayMulticastDomainGroups
=== CONT  TestAccEC2TransitGateway_serial/MulticastDomain/disappears
=== CONT  TestAccEC2TransitGateway_serial/MulticastDomain/TransitGatewayMulticastDomainAssociation
=== CONT  TestAccEC2TransitGateway_serial/MulticastDomain/basic
=== CONT  TestAccEC2TransitGateway_serial/MulticastDomain/Tags
--- PASS: TestAccEC2TransitGateway_serial (1079.27s)
    --- PASS: TestAccEC2TransitGateway_serial/MulticastDomain (0.00s)
        --- PASS: TestAccEC2TransitGateway_serial/MulticastDomain/disappears (361.15s)
        --- PASS: TestAccEC2TransitGateway_serial/MulticastDomain/basic (417.19s)
        --- PASS: TestAccEC2TransitGateway_serial/MulticastDomain/Tags (495.05s)
        --- PASS: TestAccEC2TransitGateway_serial/MulticastDomain/TransitGatewayMulticastDomainAssociation (856.38s)
        --- PASS: TestAccEC2TransitGateway_serial/MulticastDomain/TransitGatewayMulticastDomainGroups (1079.27s)
PASS
ok  	github.com/hashicorp/terraform-provider-aws/internal/service/ec2	1088.143s

File: internal/service/ec2/transit_gateway_multicast_domain_test.go [hunk 2]
```diff
@@ -0,0 +1,1007 @@
...
+		conn := acctest.Provider.Meta().(*conns.AWSClient).EC2Conn
+		id := aws.StringValue(domain.TransitGatewayMulticastDomainId)
+
+		assocSet, err := tfec2.GetTransitGatewayMulticastDomainAssociations(conn, id)
+		if err != nil {
+			return err
+		}
+
+		assocLen := len(assocSet)
+		if assocLen != count {
```
Comment (AdamTylerLynch) [inline internal/service/ec2/transit_gateway_multicast_domain_test.go:316]: Assertions should not be done in the subroutines. Assertions should be done in the `testAcc` function. Use the subroutine to fetch the data and pass it by reference to the `testAcc`.

File: internal/service/ec2/transit_gateway_multicast_domain_test.go [hunk 3]
```diff
@@ -0,0 +1,1007 @@
...
+			log.Printf("[DEBUG] Subnet IDS: %s", subnetIDs)
+			found := false
+			for _, subnetID := range subnetIDs {
+				if subnetID == actualSubnetID {
+					found = true
+					break
+				}
+			}
+
+			if !found {
```
Comment (AdamTylerLynch) [inline internal/service/ec2/transit_gateway_multicast_domain_test.go:343]: Assertions should not be done in the subroutines. Assertions should be done in the `testAcc` function. Use the subroutine to fetch the data and pass it by reference to the `testAcc`.

File: internal/service/ec2/transit_gateway_multicast_domain_test.go [hunk 4]
```diff
@@ -0,0 +1,1007 @@
...
+		id := aws.StringValue(domain.TransitGatewayMulticastDomainId)
+
+		groups, err := tfec2.SearchTransitGatewayMulticastDomainGroupsByType(conn, id, member)
+		if err != nil {
+			return err
+		}
+
+		groupLen := len(groups)
+		groupType := tfec2.ResourceTransitGatewayMulticastDomainGroupType(member)
+		if groupLen != count {
```
Comment (AdamTylerLynch) [inline internal/service/ec2/transit_gateway_multicast_domain_test.go:366]: Assertions should not be done in the subroutines. Assertions should be done in the `testAcc` function. Use the subroutine to fetch the data and pass it by reference to the `testAcc`.

File: internal/service/ec2/transit_gateway_multicast_domain_test.go [hunk 5]
```diff
@@ -0,0 +1,1007 @@
...
+resource "aws_ec2_transit_gateway_multicast_domain" "test" {
+  transit_gateway_id = aws_ec2_transit_gateway.test.id
+
+  static_source_support = "enable"
+
+  association {
+    transit_gateway_attachment_id = aws_ec2_transit_gateway_vpc_attachment.test1.id
+    subnet_ids                    = [aws_subnet.test1.id]
+  }
+  members {
```
Comment (AdamTylerLynch) [inline internal/service/ec2/transit_gateway_multicast_domain_test.go:848]: This type of multiple `members` as siblings feels unintuitive. I would rather see a members { type list }.

@ewbankkit thoughts?

The same would apply to `associations` and `sources`.

File: internal/service/ec2/transit_gateway_multicast_domain_test.go [hunk 6]
```diff
@@ -0,0 +1,1007 @@
...
+    Name = %[1]q
+  }
+}
+resource "aws_ec2_transit_gateway" "test" {
+  multicast_support = "enable"
+  tags = {
+    Name = %[1]q
+  }
+}
+resource "aws_ec2_transit_gateway_vpc_attachment" "test1" {
```
Comment (AdamTylerLynch) [inline internal/service/ec2/transit_gateway_multicast_domain_test.go:935]: I would pass in the `224.0.0.1` and `224.0.0.2` as parameters from the accTest function. Define them in the function as variables, and pass them into these `config` functions. Use those same variables for assertions in the `accTest` function.

File: website/docs/r/ec2_transit_gateway_multicast_domain.html.markdown [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+page_title: "AWS: aws_ec2_transit_gateway_multicast_domain"
+description: |-
+  Manages an EC2 Transit Gateway Multicast Domain
+---
+
+# Resource: aws_ec2_transit_gateway_multicast_domain
+
+Manages an EC2 Transit Gateway Multicast Domain.
+
+## Example Usage
```
Comment (AdamTylerLynch) [inline website/docs/r/ec2_transit_gateway_multicast_domain.html.markdown:13]: Can we include another set of larger examples that show the TGW, TGW-Multicast, Ec2 instance all working together?

https://github.com/hashicorp/terraform-provider-aws/blob/main/docs/contributing/contribution-checklists.md#documentation-update

File: website/docs/r/ec2_transit_gateway_multicast_domain.html.markdown [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+
+## Attributes Reference
+
+In addition to all arguments above, the following attributes are exported:
+
+* `id` - EC2 Transit Gateway Multicast Domain identifier
+
+## Import
+
+`aws_ec2_transit_gateway_multicast_domain` can be imported by using the EC2 Transit Gateway Multicast Domain identifier, e.g.,
```
Comment (AdamTylerLynch) [inline website/docs/r/ec2_transit_gateway_multicast_domain.html.markdown:62]: I do not see Import functionality in the resource? Can you verify `Importer:` has been defined in the resource?

**Review by AdamTylerLynch — CHANGES_REQUESTED** [review]
Over all really strong PR. 

I would like to see future PRs adhere to the branch naming conventions.
https://github.com/hashicorp/terraform-provider-aws/blob/main/docs/contributing/pullrequest-submission-and-lifecycle.md#branch-prefixes

File: internal/service/ec2/transit_gateway_multicast_domain.go [hunk 9]
```diff
@@ -0,0 +1,787 @@
...
+
+func ResourceTransitGatewayMulticastDomain() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceTransitGatewayMulticastDomainCreate,
+		Read:   resourceTransitGatewayMulticastDomainRead,
+		Update: resourceTransitGatewayMulticastDomainUpdate,
+		Delete: resourceTransitGatewayMulticastDomainDelete,
+
+		Schema: map[string]*schema.Schema{
+			"transit_gateway_id": {
```
[the remark this answers is not in the document]
Comment (NetDevAutomate) [reply internal/service/ec2/transit_gateway_multicast_domain.go]: Many thanks, I have now re-ordered schema attributes to be in alphabetical order.

File: internal/service/ec2/transit_gateway_multicast_domain.go [hunk 10]
```diff
@@ -0,0 +1,787 @@
...
+			"auto_accept_shared_associations": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Default:  ec2.AutoAcceptSharedAssociationsValueDisable,
+				ValidateFunc: validation.StringInSlice([]string{
+					ec2.AutoAcceptSharedAssociationsValueEnable,
+					ec2.AutoAcceptSharedAssociationsValueDisable,
+				}, false),
+			},
+			"igmpv2_support": {
```
[the remark this answers is not in the document]
Comment (NetDevAutomate) [reply internal/service/ec2/transit_gateway_multicast_domain.go:62]: Many thanks, this has now been added.

[the remark this answers is not in the document]
Comment (NetDevAutomate) [reply internal/service/ec2/transit_gateway_multicast_domain.go:62]: Added: ValidateFunc: validation.StringInSlice([]string{"enable", "disable"}, false),

[the remark this answers is not in the document]
Comment (NetDevAutomate) [reply internal/service/ec2/transit_gateway_multicast_domain.go:62]: Added: ValidateFunc: validation.StringInSlice([]string{"enable", "disable"}, false)

[the remark this answers is not in the document]
Comment (NetDevAutomate) [reply internal/service/ec2/transit_gateway_multicast_domain.go:62]: "creation_time": {
				Type:     schema.TypeString,
				Computed: true,
			},

File: internal/service/ec2/transit_gateway_multicast_domain.go [hunk 11]
```diff
@@ -0,0 +1,787 @@
...
+							Elem:     &schema.Schema{Type: schema.TypeString},
+							Required: true,
+							MinItems: 1,
+							Set:      schema.HashString,
+						},
+					},
+				},
+				Set: resourceTransitGatewayMulticastDomainGroupsHash,
+			},
+			"static_source_support": {
```
[the remark this answers is not in the document]
Comment (NetDevAutomate) [reply internal/service/ec2/transit_gateway_multicast_domain.go:115]: Added: ValidateFunc: validation.StringInSlice([]string{"enable", "disable"}, false),

File: internal/service/ec2/transit_gateway_multicast_domain.go [hunk 12]
```diff
@@ -0,0 +1,787 @@
...
+}
+
+func resourceTransitGatewayMulticastDomainAssociationMake(attachID string, subnetIDs *schema.Set) map[string]interface{} {
+	assoc := make(map[string]interface{})
+	assoc["transit_gateway_attachment_id"] = attachID
+	assoc["subnet_ids"] = subnetIDs
+	return assoc
+}
+
+func resourceTransitGatewayMulticastDomainGroupsReadRaw(conn *ec2.EC2, domainID string, member bool) (map[string]interface{}, error) {
```
[the remark this answers is not in the document]
Comment (NetDevAutomate) [reply internal/service/ec2/transit_gateway_multicast_domain.go:412]: Many thanks, removed.

File: internal/service/ec2/transit_gateway_multicast_domain.go [hunk 13]
```diff
@@ -0,0 +1,787 @@
...
+	if err := WaitForTransitGatewayMulticastDomainAssociation(conn, id, subnetIDs); err != nil {
+		return fmt.Errorf(
+			"error waiting for EC2 Transit Gateway Multicast Domain (%s) associations: %s", id, err)
+	}
+
+	return nil
+}
+
+func resourceTransitGatewayMulticastDomainGroupDeregister(conn *ec2.EC2, id string, groupData map[string]interface{}, member bool) error {
+	// Note: for some reason AWS made the decision to logically separate "members" from "sources" in
```
[the remark this answers is not in the document]
Comment (NetDevAutomate) [reply internal/service/ec2/transit_gateway_multicast_domain.go:603]: Corrected

File: internal/service/ec2/transit_gateway_multicast_domain.go [hunk 14]
```diff
@@ -0,0 +1,787 @@
...
+	if err := WaitForTransitGatewayMulticastDomainGroupDeregister(conn, id, groupData, member); err != nil {
+		return fmt.Errorf(
+			"error validating EC2 Transit Gateway Multicast Domain (%s) group deregistration: %s", id, err)
+	}
+
+	return nil
+}
+
+func resourceTransitGatewayMulticastDomainGroupRegister(conn *ec2.EC2, id string, groupData map[string]interface{}, member bool) error {
+	// Note: for some reason AWS made the decision to logically separate "members" from "sources" in
```
[the remark this answers is not in the document]
Comment (NetDevAutomate) [reply internal/service/ec2/transit_gateway_multicast_domain.go:642]: Corrected

File: internal/service/ec2/transit_gateway_multicast_domain.go [hunk 15]
```diff
@@ -0,0 +1,787 @@
...
+		Read:   resourceTransitGatewayMulticastDomainRead,
+		Update: resourceTransitGatewayMulticastDomainUpdate,
+		Delete: resourceTransitGatewayMulticastDomainDelete,
+
+		Schema: map[string]*schema.Schema{
+			"arn": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"association": {
```
[the remark this answers is not in the document]
Comment (NetDevAutomate) [reply internal/service/ec2/transit_gateway_multicast_domain.go:31]: I will defer to you and @ewbankkit , being able to define a multicast domain with associations in a single code block seemed to make sense as it keeps the multicast configuration contained.

File: internal/service/ec2/transit_gateway_multicast_domain.go [hunk 16]
```diff
@@ -0,0 +1,787 @@
...
+					ec2.AutoAcceptSharedAssociationsValueEnable,
+					ec2.AutoAcceptSharedAssociationsValueDisable,
+				}, false),
+			},
+			"igmpv2_support": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Default:  "disable",
+			},
+			"members": {
```
[the remark this answers is not in the document]
Comment (NetDevAutomate) [reply internal/service/ec2/transit_gateway_multicast_domain.go:67]: I will defer to you and @ewbankkit , being able to define a multicast domain with associations in a single code block seemed to make sense as it keeps the multicast configuration contained.

File: internal/service/ec2/transit_gateway_multicast_domain.go [hunk 17]
```diff
@@ -0,0 +1,787 @@
...
+						},
+					},
+				},
+				Set: resourceTransitGatewayMulticastDomainGroupsHash,
+			},
+			"owner_id": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"sources": {
```
[the remark this answers is not in the document]
Comment (NetDevAutomate) [reply internal/service/ec2/transit_gateway_multicast_domain.go:93]: I will defer to you and @ewbankkit , being able to define a multicast domain with associations in a single code block seemed to make sense as it keeps the multicast configuration contained.

File: internal/service/ec2/transit_gateway_multicast_domain_test.go [hunk 18]
```diff
@@ -0,0 +1,1007 @@
...
+		Providers:    acctest.Providers,
+		CheckDestroy: testAccCheckTransitGatewayMulticastDomainDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccTransitGatewayMulticastDomainConfigGroup1(rName),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckInstanceExists(instanceName1, &instance1),
+					testAccCheckTransitGatewayMulticastDomainExists(resourceName, &domain1),
+					testAccCheckTransitGatewayMulticastDomainGroups(&domain1, 1, true, map[string][]*ec2.Instance{
+						"224.0.0.1": {&instance1},
```
[the remark this answers is not in the document]
Comment (NetDevAutomate) [reply internal/service/ec2/transit_gateway_multicast_domain_test.go:178]: Done

File: internal/service/ec2/transit_gateway_multicast_domain_test.go (hunk 6 above)
Comment (NetDevAutomate) [reply internal/service/ec2/transit_gateway_multicast_domain_test.go:935]: Done

File: website/docs/r/ec2_transit_gateway_multicast_domain.html.markdown (hunk 7 above)
Comment (NetDevAutomate) [reply website/docs/r/ec2_transit_gateway_multicast_domain.html.markdown:13]: Added a full example

File: website/docs/r/ec2_transit_gateway_multicast_domain.html.markdown (hunk 8 above)
Comment (NetDevAutomate) [reply website/docs/r/ec2_transit_gateway_multicast_domain.html.markdown:62]: Added, many thanks for catching this.

File: internal/service/ec2/transit_gateway_multicast_domain_data_source.go [hunk 19]
```diff
@@ -0,0 +1,189 @@
...
+			"auto_accept_shared_associations": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Default:  ec2.AutoAcceptSharedAssociationsValueDisable,
+				ValidateFunc: validation.StringInSlice([]string{
+					ec2.AutoAcceptSharedAssociationsValueEnable,
+					ec2.AutoAcceptSharedAssociationsValueDisable,
+				}, false),
+			},
+			"filter": DataSourceFiltersSchema(),
```
[the remark this answers is not in the document]
Comment (NetDevAutomate) [reply internal/service/ec2/transit_gateway_multicast_domain_data_source.go:56]: Added

File: internal/service/ec2/transit_gateway_multicast_domain_data_source.go [hunk 20]
```diff
@@ -0,0 +1,189 @@
...
+						},
+					},
+				},
+				Set: resourceTransitGatewayMulticastDomainGroupsHash,
+			},
+			"owner_id": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"sources": {
```
[the remark this answers is not in the document]
Comment (NetDevAutomate) [reply internal/service/ec2/transit_gateway_multicast_domain_data_source.go:93]: Added

File: internal/service/ec2/transit_gateway_multicast_domain.go [hunk 21]
```diff
@@ -0,0 +1,787 @@
...
+	tftags "github.com/hashicorp/terraform-provider-aws/internal/tags"
+)
+
+func ResourceTransitGatewayMulticastDomain() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceTransitGatewayMulticastDomainCreate,
+		Read:   resourceTransitGatewayMulticastDomainRead,
+		Update: resourceTransitGatewayMulticastDomainUpdate,
+		Delete: resourceTransitGatewayMulticastDomainDelete,
+
```
[the remark this answers is not in the document]
Comment (NetDevAutomate) [reply internal/service/ec2/transit_gateway_multicast_domain.go:25]: Added, many thanks for catching this.

**Review by AdamTylerLynch — APPROVED** [review]
LGTM 🚀

**Comment by ewbankkit** [comment]
Let's go with the AWS CloudFormation model here and have additional resources:

* `aws_ec2_transit_gateway_multicast_domain_association` inspired by [`AWS::EC2::TransitGatewayMulticastDomainAssociation`](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ec2-transitgatewaymulticastdomainassociation.html)
* `aws_ec2_transit_gateway_multicast_group_member` inspired by [`AWS::EC2::TransitGatewayMulticastGroupMember`](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ec2-transitgatewaymulticastgroupmember.html)
* `aws_ec2_transit_gateway_multicast_group_source` inspired by [`AWS::EC2::TransitGatewayMulticastGroupSource`](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ec2-transitgatewaymulticastgroupsource.html)

**Comment by NetDevAutomate** [comment]
> Let's go with the AWS CloudFormation model here and have additional resources:
> 
> * `aws_ec2_transit_gateway_multicast_domain_association` inspired by [`AWS::EC2::TransitGatewayMulticastDomainAssociation`](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ec2-transitgatewaymulticastdomainassociation.html)
> * `aws_ec2_transit_gateway_multicast_group_member` inspired by [`AWS::EC2::TransitGatewayMulticastGroupMember`](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ec2-transitgatewaymulticastgroupmember.html)
> * `aws_ec2_transit_gateway_multicast_group_source` inspired by [`AWS::EC2::TransitGatewayMulticastGroupSource`](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ec2-transitgatewaymulticastgroupsource.html)

@ewbankkit Many thanks, mirroring the CloudFormation model makes a lot of sense.

**Comment by ewbankkit** [comment]
@NetDevAutomate @AdamTylerLynch @drewmullen For a follow up should we add a `aws_ec2_transit_gateway_multicast_domain_association_accepter` resource for those domains where `auto_accept_shared_associations = "disable"`?

**Review by ewbankkit — APPROVED** [review]
LGTM :rocket:.

##### Commercial

```console
% make testacc TESTS=TestAccEC2TransitGatewayDataSource_serial/Gateway/Filter PKG=ec2
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/ec2/... -v -count 1 -parallel 20 -run='TestAccEC2TransitGatewayDataSource_serial/Gateway/Filter'  -timeout 180m
=== RUN   TestAccEC2TransitGatewayDataSource_serial
=== RUN   TestAccEC2TransitGatewayDataSource_serial/DxGatewayAttachment
=== RUN   TestAccEC2TransitGatewayDataSource_serial/DxGatewayAttachment/Filter
=== RUN   TestAccEC2TransitGatewayDataSource_serial/Gateway
=== RUN   TestAccEC2TransitGatewayDataSource_serial/Gateway/Filter
--- PASS: TestAccEC2TransitGatewayDataSource_serial (1648.95s)
    --- PASS: TestAccEC2TransitGatewayDataSource_serial/DxGatewayAttachment (1413.49s)
        --- PASS: TestAccEC2TransitGatewayDataSource_serial/DxGatewayAttachment/Filter (1413.49s)
    --- PASS: TestAccEC2TransitGatewayDataSource_serial/Gateway (235.46s)
        --- PASS: TestAccEC2TransitGatewayDataSource_serial/Gateway/Filter (235.46s)
PASS
ok  	github.com/hashicorp/terraform-provider-aws/internal/service/ec2	1652.703s
% make testacc TESTS=TestAccEC2TransitGatewayDataSource_serial/MulticastDomain PKG=ec2
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/ec2/... -v -count 1 -parallel 20 -run='TestAccEC2TransitGatewayDataSource_serial/MulticastDomain'  -timeout 180m
=== RUN   TestAccEC2TransitGatewayDataSource_serial
=== RUN   TestAccEC2TransitGatewayDataSource_serial/MulticastDomain
=== RUN   TestAccEC2TransitGatewayDataSource_serial/MulticastDomain/Filter
=== RUN   TestAccEC2TransitGatewayDataSource_serial/MulticastDomain/ID
--- PASS: TestAccEC2TransitGatewayDataSource_serial (859.82s)
    --- PASS: TestAccEC2TransitGatewayDataSource_serial/MulticastDomain (859.81s)
        --- PASS: TestAccEC2TransitGatewayDataSource_serial/MulticastDomain/Filter (378.79s)
        --- PASS: TestAccEC2TransitGatewayDataSource_serial/MulticastDomain/ID (481.03s)
PASS
ok  	github.com/hashicorp/terraform-provider-aws/internal/service/ec2	863.406s
% make testacc TESTS=TestAccEC2TransitGateway_serial/Gateway/basic PKG=ec2
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/ec2/... -v -count 1 -parallel 20 -run='TestAccEC2TransitGateway_serial/Gateway/basic'  -timeout 180m
=== RUN   TestAccEC2TransitGateway_serial
=== RUN   TestAccEC2TransitGateway_serial/Gateway
=== RUN   TestAccEC2TransitGateway_serial/Gateway/basic
--- PASS: TestAccEC2TransitGateway_serial (216.96s)
    --- PASS: TestAccEC2TransitGateway_serial/Gateway (216.96s)
        --- PASS: TestAccEC2TransitGateway_serial/Gateway/basic (216.96s)
PASS
ok  	github.com/hashicorp/terraform-provider-aws/internal/service/ec2	220.887s
% make testacc TESTS=TestAccEC2TransitGateway_serial/MulticastDomain PKG=ec2
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/ec2/... -v -count 1 -parallel 20 -run='TestAccEC2TransitGateway_serial/MulticastDomain'  -timeout 180m
=== RUN   TestAccEC2TransitGateway_serial
=== RUN   TestAccEC2TransitGateway_serial/MulticastDomainAssociation
=== RUN   TestAccEC2TransitGateway_serial/MulticastDomainAssociation/basic
=== RUN   TestAccEC2TransitGateway_serial/MulticastDomainAssociation/disappears
=== RUN   TestAccEC2TransitGateway_serial/MulticastDomainAssociation/DomainDisappears
=== RUN   TestAccEC2TransitGateway_serial/MulticastDomainAssociation/TwoAssociations
=== RUN   TestAccEC2TransitGateway_serial/MulticastDomain
=== RUN   TestAccEC2TransitGateway_serial/MulticastDomain/basic
=== RUN   TestAccEC2TransitGateway_serial/MulticastDomain/disappears
=== RUN   TestAccEC2TransitGateway_serial/MulticastDomain/tags
=== RUN   TestAccEC2TransitGateway_serial/MulticastDomain/IGMPv2Support
--- PASS: TestAccEC2TransitGateway_serial (3435.81s)
    --- PASS: TestAccEC2TransitGateway_serial/MulticastDomainAssociation (2004.23s)
        --- PASS: TestAccEC2TransitGateway_serial/MulticastDomainAssociation/basic (482.98s)
        --- PASS: TestAccEC2TransitGateway_serial/MulticastDomainAssociation/disappears (503.32s)
        --- PASS: TestAccEC2TransitGateway_serial/MulticastDomainAssociation/DomainDisappears (536.22s)
        --- PASS: TestAccEC2TransitGateway_serial/MulticastDomainAssociation/TwoAssociations (481.72s)
    --- PASS: TestAccEC2TransitGateway_serial/MulticastDomain (1431.57s)
        --- PASS: TestAccEC2TransitGateway_serial/MulticastDomain/basic (336.97s)
        --- PASS: TestAccEC2TransitGateway_serial/MulticastDomain/disappears (317.72s)
        --- PASS: TestAccEC2TransitGateway_serial/MulticastDomain/tags (359.24s)
        --- PASS: TestAccEC2TransitGateway_serial/MulticastDomain/IGMPv2Support (417.65s)
PASS
ok  	github.com/hashicorp/terraform-provider-aws/internal/service/ec2	3439.360s
% make testacc TESTS=TestAccEC2TransitGateway_serial/MulticastGroup PKG=ec2
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/ec2/... -v -count 1 -parallel 20 -run='TestAccEC2TransitGateway_serial/MulticastGroup'  -timeout 180m
=== RUN   TestAccEC2TransitGateway_serial
=== RUN   TestAccEC2TransitGateway_serial/MulticastGroupMember
=== RUN   TestAccEC2TransitGateway_serial/MulticastGroupMember/basic
=== RUN   TestAccEC2TransitGateway_serial/MulticastGroupMember/disappears
=== RUN   TestAccEC2TransitGateway_serial/MulticastGroupMember/DomainDisappears
=== RUN   TestAccEC2TransitGateway_serial/MulticastGroupMember/TwoMembers
=== RUN   TestAccEC2TransitGateway_serial/MulticastGroupSource
=== RUN   TestAccEC2TransitGateway_serial/MulticastGroupSource/disappears
=== RUN   TestAccEC2TransitGateway_serial/MulticastGroupSource/DomainDisappears
=== RUN   TestAccEC2TransitGateway_serial/MulticastGroupSource/basic
--- PASS: TestAccEC2TransitGateway_serial (3481.58s)
    --- PASS: TestAccEC2TransitGateway_serial/MulticastGroupMember (2000.02s)
        --- PASS: TestAccEC2TransitGateway_serial/MulticastGroupMember/basic (437.51s)
        --- PASS: TestAccEC2TransitGateway_serial/MulticastGroupMember/disappears (486.55s)
        --- PASS: TestAccEC2TransitGateway_serial/MulticastGroupMember/DomainDisappears (594.82s)
        --- PASS: TestAccEC2TransitGateway_serial/MulticastGroupMember/TwoMembers (481.14s)
    --- PASS: TestAccEC2TransitGateway_serial/MulticastGroupSource (1481.56s)
        --- PASS: TestAccEC2TransitGateway_serial/MulticastGroupSource/disappears (415.15s)
        --- PASS: TestAccEC2TransitGateway_serial/MulticastGroupSource/DomainDisappears (585.59s)
        --- PASS: TestAccEC2TransitGateway_serial/MulticastGroupSource/basic (480.82s)
PASS
ok  	github.com/hashicorp/terraform-provider-aws/internal/service/ec2	3485.016s
```

##### GovCloud

```console

```

**Comment by ewbankkit** [comment]
@NetDevAutomate Thanks for the contribution :tada: :clap:.

**Comment by NetDevAutomate** [comment]
> @NetDevAutomate Thanks for the contribution 🎉 👏.

@ewbankkit Thank you so much for your support, it really is greatly appreciated.

**Pull request merged**